pylint/checkers/utils.py

```diff
@@ -1361,45 +1361,95 @@ def safe_infer(
     If compare_constants is True and if multiple constants are inferred,
     unequal inferred values are also considered ambiguous and return None.
     """
-    inferred_types: set[str | None] = set()
+    infer_gen = node.infer(context=context)
+    first_infer = _safe_infer_validate_first(infer_gen)
     try:
-        infer_gen = node.infer(context=context)
-        value = next(infer_gen)
+        for inferred in infer_gen:
+            is_same_type = _safe_infer_compare_to_first(
+                first_infer, inferred, compare_constants
+            )
+            if not is_same_type:
+                return None
     except astroid.InferenceError:
-        return None
+        return None  # There is some kind of ambiguity
+    except StopIteration:
+        return first_infer
     except Exception as e:  # pragma: no cover
         raise AstroidError from e
+    return first_infer
 
-    if value is not astroid.Uninferable:
-        inferred_types.add(_get_python_type_of_node(value))
 
+@lru_cache(maxsize=1024)
+def safe_infer_multiple(
+    node: nodes.NodeNG,
+    context: InferenceContext | None = None,
+    *,
+    compare_constants: bool = False,
+) -> Any | None:
```

Comment:
Why is this `Any`?

Reply:
I asked myself the same question. But this is what `mypy`  accepted. :S

Reply:
Yes, that's because `mypy` doesn't do any checking on `astroid`. It should be `InferenceResult`.